File: 3575.maximum-good-subtree-score.go
#
# @lc app=leetcode id=3575 lang=golang
#
# [3575] Maximum Good Subtree Score
#

# @lc code=start
func goodSubtreeSum(vals []int, par []int) int {
    n := len(vals)
    adj := make([][]int, n)
    for i := 1; i < n; i++ {
        adj[par[i]] = append(adj[par[i]], i)
    }
    mask := make([]int, n)
    valid := make([]bool, n)
    for i := 0; i < n; i++ {
        temp := vals[i]
        seen := 0
        isValid := true
        for temp > 0 {
            d := temp % 10
            bit := 1 << d
            if (seen & bit) != 0 {
                isValid = false
                break
            }
            seen |= bit
            temp /= 10
        }
        valid[i] = isValid
        if isValid {
            mask[i] = seen
        }
    }
    const SZ = 1 << 10
    const MOD int64 = 1000000007
    maxScores := make([]int64, n)
    var dfs func(int) []int64
    dfs = func(u int) []int64 {
        descendants := make([]int64, SZ)
        for i := range descendants {
            descendants[i] = -1
        }
        descendants[0] = 0
        for _, v := range adj[u] {
            child := dfs(v)
            newDesc := make([]int64, SZ)
            for i := range newDesc {
                newDesc[i] = -1
            }
            currActive := make([]int, 0, SZ)
            for m := 0; m < SZ; m++ {
                if descendants[m] != -1 {
                    currActive = append(currActive, m)
                }
            }
            childActive := make([]int, 0, SZ)
            for m := 0; m < SZ; m++ {
                if child[m] != -1 {
                    childActive = append(childActive, m)
                }
            }
            for _, m1 := range currActive {
                s1 := descendants[m1]
                for _, m2 := range childActive {
                    s2 := child[m2]
                    if (m1 & m2) == 0 {
                        nm := m1 | m2
                        ns := s1 + s2
                        if newDesc[nm] < ns {
                            newDesc[nm] = ns
                        }
                    }
                }
            }
            descendants = newDesc
        }
        descActive := make([]int, 0, SZ)
        for m := 0; m < SZ; m++ {
            if descendants[m] != -1 {
                descActive = append(descActive, m)
            }
        }
        subtree := make([]int64, SZ)
        for i := range subtree {
            subtree[i] = -1
        }
        // without u
        for _, m := range descActive {
            subtree[m] = descendants[m]
        }
        // with u
        if valid[u] {
            mu := mask[u]
            vu := int64(vals[u])
            for _, m := range descActive {
                if (m & mu) == 0 {
                    nm := m | mu
                    ns := descendants[m] + vu
                    if subtree[nm] < ns {
                        subtree[nm] = ns
                    }
                }
            }
        }
        var mx int64 = 0
        for m := 0; m < SZ; m++ {
            if subtree[m] > mx {
                mx = subtree[m]
            }
        }
        maxScores[u] = mx
        return subtree
    }
    dfs(0)
    var ans int64 = 0
    for _, m := range maxScores {
        ans = (ans + m) % MOD
    }
    return int(ans)
}
# @lc code=end

File: 3575.maximum-good-subtree-score.java
#
# @lc app=leetcode id=3575 lang=java
#
# [3575] Maximum Good Subtree Score
#

# @lc code=start
class Solution {
    public int goodSubtreeSum(int[] vals, int[] par) {
        final int MOD = 1000000007;
        int n = vals.length;
        int[] degree = new int[n];
        for (int i = 1; i < n; i++) {
            degree[par[i]]++;
        }
        int[][] chld = new int[n][];
        int[] ptr = new int[n];
        for (int i = 0; i < n; i++) {
            chld[i] = new int[degree[i]];
        }
        for (int i = 1; i < n; i++) {
            int p = par[i];
            chld[p][ptr[p]++] = i;
        }
        int[] masks = new int[n];
        long[] nodevals = new long[n];
        for (int i = 0; i < n; i++) {
            String s = Integer.toString(vals[i]);
            boolean[] digUsed = new boolean[10];
            boolean dup = false;
            int msk = 0;
            for (int k = 0; k < s.length(); k++) {
                int d = s.charAt(k) - '0';
                if (digUsed[d]) {
                    dup = true;
                    break;
                }
                digUsed[d] = true;
                msk |= (1 << d);
            }
            if (dup) {
                nodevals[i] = 0L;
                masks[i] = 0;
            } else {
                nodevals[i] = vals[i];
                masks[i] = msk;
            }
        }
        long[] maxScore = new long[n];
        int MS = 1 << 10;
        class DfsHelper {
            long[] dfs(int u, int[] masks, long[] nodevals, int[][] chld, long[] maxScore, int MS) {
                long[] curr = new long[MS];
                for (int i = 0; i < MS; i++) {
                    curr[i] = -1L;
                }
                curr[0] = 0L;
                for (int ci = 0; ci < chld[u].length; ci++) {
                    int v = chld[u][ci];
                    long[] chdp = dfs(v, masks, nodevals, chld, maxScore, MS);
                    long[] nextt = new long[MS];
                    for (int i = 0; i < MS; i++) {
                        nextt[i] = -1L;
                    }
                    java.util.ArrayList<java.lang.Integer> currValid = new java.util.ArrayList<java.lang.Integer>();
                    for (int m1 = 0; m1 < MS; m1++) {
                        if (curr[m1] >= 0L) {
                            currValid.add(m1);
                        }
                    }
                    java.util.ArrayList<java.lang.Integer> chValid = new java.util.ArrayList<java.lang.Integer>();
                    for (int m2 = 0; m2 < MS; m2++) {
                        if (chdp[m2] >= 0L) {
                            chValid.add(m2);
                        }
                    }
                    long est = (long) currValid.size() * chValid.size();
                    if (est > 500000L) {
                        // full loop
                        for (int m1 = 0; m1 < MS; m1++) {
                            if (curr[m1] < 0L) continue;
                            for (int m2 = 0; m2 < MS; m2++) {
                                if (chdp[m2] < 0L) continue;
                                if ((m1 & m2) != 0) continue;
                                int nm = m1 | m2;
                                long ns = curr[m1] + chdp[m2];
                                if (nextt[nm] < ns) {
                                    nextt[nm] = ns;
                                }
                            }
                        }
                    } else {
                        // list loop
                        for (int j = 0; j < currValid.size(); j++) {
                            int m1 = currValid.get(j);
                            for (int k = 0; k < chValid.size(); k++) {
                                int m2 = chValid.get(k);
                                if ((m1 & m2) == 0) {
                                    int nm = m1 | m2;
                                    long ns = curr[m1] + chdp[m2];
                                    if (nextt[nm] < ns) {
                                        nextt[nm] = ns;
                                    }
                                }
                            }
                        }
                    }
                    curr = nextt;
                }
                long maxU = 0L;
                for (int m = 0; m < MS; m++) {
                    if (curr[m] > maxU) {
                        maxU = curr[m];
                    }
                }
                int omask = masks[u];
                long oval = nodevals[u];
                long maxAddU = 0L;
                for (int m = 0; m < MS; m++) {
                    if (curr[m] < 0L) continue;
                    if ((m & omask) == 0) {
                        maxAddU = Math.max(maxAddU, curr[m] + oval);
                    }
                }
                maxU = Math.max(maxU, maxAddU);
                maxScore[u] = maxU;
                for (int m = 0; m < MS; m++) {
                    if (curr[m] < 0L) continue;
                    if ((m & omask) == 0) {
                        int nm = m | omask;
                        long ns = curr[m] + oval;
                        if (curr[nm] < ns) {
                            curr[nm] = ns;
                        }
                    }
                }
                return curr;
            }
        }
        DfsHelper helper = new DfsHelper();
        helper.dfs(0, masks, nodevals, chld, maxScore, MS);
        long ans = 0L;
        for (long x : maxScore) {
            ans = (ans + x) % MOD;
        }
        return (int) ans;
    }
}
# @lc code=end